mentat/auto_context.py

```diff
@@ -0,0 +1,150 @@
+import json
+from textwrap import dedent
+from typing import Optional, cast
+
+import openai
+
+from mentat.code_feature import CodeFeature, CodeMessageLevel
+from mentat.llm_api import count_tokens
+from mentat.session_context import SESSION_CONTEXT
+
+"""
+Context selection is similar to the Knapsack Problem: given a set of items (features),
+and a knapsack (context size), choose the best subset of items that fit. In our case we
+can also select which level include. We implement two methods:
+
+- Greedy: add features one-by-one, at the most detailed level possible, until the maximum
+    token limit is reached.
+
+- LLM: create a full-context of code using the greedy method, then ask the LLM to select
+    only the relevant features. Return its output, up to the token limit.
+"""
+
+
+def select_features_greedy(
+    features: list[CodeFeature],
+    max_tokens: int,
+    model: str = "gpt-4",
+    levels: list[CodeMessageLevel] = [],
+) -> list[CodeFeature]:
+    """Use the greedy method to return a subset of features max_tokens long"""
+    output = list[CodeFeature]()
+    remaining_tokens = max_tokens
+    for feature in features:
+        _levels = list(set(levels) | {feature.level})
+        _levels = sorted(list(_levels), key=lambda v: v.rank)
+        for level in _levels:
+            feature.level = level
+            if feature.count_tokens(model) <= remaining_tokens:
+                output.append(feature)
+                remaining_tokens -= feature.count_tokens(model)
+                break
```

Comment:
just to confirm, the result of this we'll get a bunch of features at the highest level of detail, and then probably just a couple features at a lower level? And then nothing for remaining features?

Reply:
Correct. Kindof a depth-first selection instead of our previous breadth-first.